google-cloud-firestore/src/test/java/com/google/cloud/firestore/BulkWriterTest.java

```diff
@@ -1119,11 +1118,10 @@ public void failsWritesAfterAllRetryAttemptsFail() throws Exception {
         };
 
     doAnswer(
-            (Answer<ApiFuture<GeneratedMessageV3>>)
-                mock -> {
-                  retryAttempts[0]++;
-                  return RETRYABLE_FAILED_FUTURE;
-                })
+            mock -> {
```

Comment:
will this test the new changes?

Reply:
No, it's hard to test these changes. We could mock the backend stream, but then we are essentially only testing the behavior of the mock. If you know of a pre-existing implementation/fake of a GRPC stream that we can use to test this behavior, then we can add a test. A homegrown implementation that validates that our code follows our ow assumptions will not provide us with much meaningful test coverage.

Reply:
Would something like this be helpful? https://github.com/googleapis/java-bigtable/blob/main/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/gaxx/testing/FakeStreamingApi.java
Examples in tests:
https://github.com/googleapis/java-bigtable/blob/main/google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/readrows/FilterMarkerRowsCallableTest.java#L38

(It's using ApiStreamObserver, but I imagine could be refactored to this use case)

Reply:
Added test queryWatchShutsDownStreamOnPermissionDenied() that re-uses some of the existing functionality.